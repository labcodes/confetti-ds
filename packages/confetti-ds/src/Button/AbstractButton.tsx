import React, { SyntheticEvent } from "react";
import { isUndefined } from "lodash";

import Icon from "../Icon";
import { IconTypes } from "../constants";
import { BaseButtonProps } from "./types";

interface AbstractButtonProps extends BaseButtonProps {
  /** Button style variation. */
  kind?: "default" | "outline" | "text";
  /** Sets the icon related to the button label. Default state: no icon. */
  theme?:
    | "teal"
    | "purple"
    | "light"
    | "dark"
    | "warning"
    | "destructive"
    | "confirmation"
    | "warning-invert"
    | "destructive-invert"
    | "confirmation-invert";
}

export default function AbstractButton({
  label,
  type = "button" as const,
  kind,
  icon,
  theme = "teal",
  size = "small",
  ariaDisabled = false,
  disabled = false,
  onClick = (event) => {},
  fullWidth = false,
  tabIndex,
  hasIcon = false,
}: AbstractButtonProps) {
  const renderIcon = () =>
    icon ? (
      <Icon type={icon} color="white" className="lab-btn__icon" />
    ) : (
      <Icon type={"Add"} color="white" className="lab-btn__icon" />
    );
  /*
  Set icon color to be the icon color of default button.
  Icon color changes according to button type and exception on _buttons.scss
  */

  const handleOnClick = (event) => {
    if (!isUndefined(onClick)) {
      onClick(event);
    }
  };
  return (
    <button
      type={type}
      size={size}
      kind={kind}
      className={
        `lab-btn` +
        ` lab-btn--${kind} lab-btn--${size}` +
        `${theme ? ` lab-btn--${theme}` : ""}` +
        `${fullWidth ? ` lab-btn--block` : ""}`
      }
      onClick={!ariaDisabled ? handleOnClick : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
      {...(tabIndex ? { tabIndex } : undefined)}
    >
      {hasIcon ? renderIcon() : null}
      {label}
    </button>
  );
}
