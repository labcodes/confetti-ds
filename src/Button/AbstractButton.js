import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";

import Icon from "../Icon";
import { ICON_TYPES } from "../constants";

export default class AbstractButton extends React.Component {
  static propTypes = {
    /** HTML type of the Button. */
    type: PropTypes.oneOf(["button", "submit", "reset"]),
    /** Button style variation. */
    variant: PropTypes.oneOf(["default", "outline", "text"]),
    /** This is the button label. */
    text: PropTypes.string.isRequired,
    /** Sets the icon related to the button label. Default state: no icon. */
    icon: PropTypes.oneOf(ICON_TYPES),
    /** Sets a special color skin to the button. */
    skin: PropTypes.oneOf([
      "",
      "light",
      "dark",
      "warning",
      "destructive",
      "warning-invert",
      "destructive-invert",
      "confirmation-invert",
    ]),
    /** Sets the button's height. Small = 32px, Normal = 40px, Large = 48px. */
    size: PropTypes.oneOf(["normal", "small", "large"]),
    /** Disables the Button. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Button. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Action to be executed when the button is clicked. */
    onClick: PropTypes.func,
    /** Makes the button expand to its container's full width. */
    fullWidth: PropTypes.bool,
    /* Sets the order of which the elements will be focused on. Default value: 0. */
    tabIndex: PropTypes.string,
  };

  static defaultProps = {
    type: "button",
    variant: "default",
    skin: "",
    icon: undefined,
    size: "normal",
    disabled: false,
    ariaDisabled: false,
    onClick: () => {},
    fullWidth: false,
    tabIndex: "0",
  };

  icon = () => {
    const { icon } = this.props;

    /*
    Set icon color to be the icon color of default button.
    Icon color changes according to button type and exception on _buttons.scss
    */
    return icon ? (
      <Icon type={icon} color="white" size="petit" className="lab-btn__icon" />
    ) : (
      ""
    );
  };

  handleOnClick = (event) => {
    const { onClick } = this.props;
    if (!isUndefined(onClick)) {
      onClick(event);
    }
  };

  render() {
    const {
      type,
      text,
      variant,
      skin,
      size,
      disabled,
      fullWidth,
      ariaDisabled,
      tabIndex,
    } = this.props;
    return (
      <button
        // eslint-disable-next-line react/button-has-type
        type={type}
        className={
          `lab-btn` +
          ` lab-btn--${variant} lab-btn--${size}` +
          `${skin ? ` lab-btn--${skin}` : ""}` +
          `${fullWidth ? ` lab-btn--block` : ""}`
        }
        onClick={!ariaDisabled ? this.handleOnClick : () => {}}
        disabled={(!ariaDisabled && disabled) || undefined}
        aria-disabled={ariaDisabled || undefined}
        tabIndex={tabIndex}
      >
        {this.icon()}
        {text}
      </button>
    );
  }
}
