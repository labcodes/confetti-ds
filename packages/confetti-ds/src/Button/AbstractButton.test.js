import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import AbstractButton from "./AbstractButton";

describe("AbstractButton", () => {
  it("renders with base props for default variant", async () => {
    const renderedComponent = renderer
      .create(<AbstractButton variant="default" text="Test Default Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with base props for outline variant", async () => {
    const renderedComponent = renderer
      .create(<AbstractButton variant="outline" text="Test Outline Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with base props for text variant", async () => {
    const renderedComponent = renderer
      .create(<AbstractButton variant="text" text="Test Text Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("doesn't call onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const shallowComponent = shallow(
      <AbstractButton
        ariaDisabled
        onClick={mockOnClick}
        variant="text"
        text="Test Text Button"
      />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    shallowComponent.find("button").simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(0);
  });
  it("renders as expected when tabIndex is passed", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractButton
          variant="default"
          text="Test tabIndex"
          // eslint-disable-next-line jsx-a11y/tabindex-no-positive
          tabIndex="1"
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
