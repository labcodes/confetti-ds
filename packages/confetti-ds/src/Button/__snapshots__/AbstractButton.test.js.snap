// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractButton renders with base props for default variant 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Default Button
</button>
`;

exports[`AbstractButton renders with base props for outline variant 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Outline Button
</button>
`;

exports[`AbstractButton renders with base props for text variant 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Text Button
</button>
`;
